File: java/LinkedListCycle.java
/*
 * Filename: LinkedListCycle.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Sun 05 Jul 2015 05:11:18 PM EDT
 */

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
	public boolean hasCycle(ListNode head) {
		ListNode p1 = head;
		ListNode p2 = head;

		while (p1 != null && p1.next != null) {
			p1 = p1.next.next;
			p2 = p2.next;

			if (p1 == p2) {
				return true;
			}
		}

		return false;
	}
}

File: cpp/linkedListCycle.cpp
/*
 * Filename: linkedListCycle.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Thu 14 Jan 2016 09:41:14 PM EST
 * Description: 
 */

#include <iostream>

using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution
{
	public:
		bool hasCycle(ListNode *head)
		{
			ListNode* p = head;
			ListNode* q = head;

			while (p && p->next)
			{
				p = p->next->next;
				q = q->next;

				if (p == q)
				{
					return true;
				}
			}

			return false;
		}
};
